vopr: allow fast writes

matklad opened the pull request:
In the current VOPR config, the minimum mean storage latency is 3 ticks,
which is 30 ms, which is just ridiculous. We generally expect almost all
IOPs to complete faster than in a single tick.

To model "faster than a tick" realistically, I change the simulator to
drain storage&network IO "within" each tick. As there could be multiple
immediately ready IOPs, the simulator randomizes the order in which
completions are invoked.

matklad committed:
vopr: allow fast writes

In the current VOPR config, the minimum mean storage latency is 3 ticks,
which is 30 ms, which is just ridiculous. We generally expect almost all
IOPs to complete faster than in a single tick.

To model "faster than a tick" realistically, I change the simulator to
drain storage&network IO "within" each tick. As there could be multiple
immediately ready IOPs, the simulator randomizes the order in which
completions are invoked.

matklad committed:
io: split tick into tick and run

We are overloading `fn tick`. Sometmes we use it as a passive function
to notify about passage of time (replica.tick), sometimes we use it as
an active driver of progress (storage.tick).

Rename the second usage to `run`, which rhymes with our existing run:

    pub fn run(self: *IO) !void
    pub fn run_for_ns(self: *IO, nanoseconds: u63) !void

sentientwaffle commented on src/testing/cluster.zig:
```diff
@@ -446,6 +446,29 @@ pub fn ClusterType(comptime StateMachineType: anytype) type {
         }
 
         pub fn tick(cluster: *Cluster) void {
+            // Interleave storage and network progres, to allow for faster-than-a-tick IO.
```
```suggestion
            // Interleave storage and network progress, to allow for faster-than-a-tick IO.
```

sentientwaffle commented on src/testing/storage.zig:
```diff
@@ -335,25 +335,60 @@ pub const Storage = struct {
 
-    pub fn tick(storage: *Storage) void {
-        storage.ticks += 1;
+    pub fn progress(storage: *Storage) bool {
```
Does this name use "progress" as a noun or a verb? i.e. "Is there progress?" vs "Go make progress.". It sort of seems like both at the same time right now.

matklad replied:
A verb, progrEss  (I think both your example are nouns though). 

It is ambiguous, I’ll think about a better name, and I am open for suggestions.

sentientwaffle replied:
Ah yeah, both of my examples were nouns. Maybe `forward`? Though in the context of the network that might be confusing still. Maybe "step", though that is pretty close to "tick". "advance"?

matklad replied:
Ah, step would be perfect, as this _is_ a counter-part of `run` from the next commit!

matklad replied:
Yeah, I love this:

```zig
    pub fn run(storage: *Storage) void {
        while (storage.step()) {}
        storage.tick();
    }
```

jorangreef replied:
"advance" was also just what I was thinking—and then @sentientwaffle voiced that, and "step" looks even better!

sentientwaffle commented on src/testing/cluster.zig:
There is a CFO seed, which can be fixed by:

```diff
@@ -843,6 +859,11 @@ pub fn ClusterType(comptime StateMachineType: anytype) type {
                 },
                 .sync_stage_changed => switch (replica.syncing) {
                     .idle => cluster.log_replica(.sync, replica.replica),
+                    .updating_checkpoint => {
+                        cluster.state_checker.check_state(replica.replica) catch |err| {
+                            fatal(.correctness, "state checker error: {}", .{err});
+                        };
+                    },
                     else => {},
                 },
                 .client_evicted => |client_id| cluster.cluster_on_eviction(client_id),
```

I think this PR causes state sync to finish fast enough in some cases that the state checker wasn't ever registering it, so it failed to capture the op jump here:

```zig
        pub fn check_state(state_checker: *StateChecker, replica_index: u8) !void {
            const replica = &state_checker.replicas[replica_index];
            if (replica.syncing == .updating_checkpoint) {
                // Allow a syncing replica to fast-forward its commit.
                //
                // But "fast-forwarding" may actually move commit_min slightly backwards:
                // 1. Suppose op X is a checkpoint trigger.
                // 2. We are committing op X-1 but are stuck due to a block that does not exist in
                //    the cluster anymore.
                // 3. When we sync, `commit_min` "backtracks", to `X - lsm_compaction_ops`.
                const commit_min_source = state_checker.commit_mins[replica_index];
                const commit_min_target =
                    replica.syncing.updating_checkpoint.header.op;
                assert(commit_min_source <= commit_min_target + constants.lsm_compaction_ops);
                state_checker.commit_mins[replica_index] = commit_min_target;
                return;
            }
```

matklad committed:
vopr: fix false positive

This was uncovered by fast IO, which made state checker miss new
commit_min.

Seed: ./zig/zig build -Drelease vopr -- 1935672942318629811

matklad commented:
A new seed, haven't looked into it yet: `./zig/zig build -Drelease vopr -- 1559163846617790652`

matklad committed:
vopr: fix packet leak

Need to eagerly deinit the packets, otherwise we run out of message bus
messages.

Seed: ./zig/zig build vopr -- 1559163846617790652

matklad commented:
Green at 338 273!
